### PR TITLE
Allow custom comments to Phabricator

### PR DESCRIPTION
**Pull request by sectioneight**
https://code.uberinternal.com/T50253

### CODE REVIEW

**Comment by sectioneight** [comment]
image
image

**Comment by sectioneight** [comment]
@jjx

File: src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java [hunk 1]
```diff
@@ -178,23 +180,37 @@ public final boolean perform(final AbstractBuild<?, ?> build, final Launcher lau
...
 
         diff.setBuildFinished(build.getResult());
+        String customComment = null;
+        try {
+            customComment = getRemoteComment(build, logger, this.commentFile);
+        } catch(InterruptedException e) {
+            e.printStackTrace(logger);
+        } catch (IOException e) {
+            Util.displayIOException(e, listener);
+
```
Comment (jjx) [inline src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java]: Remove this extra newline

File: src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java [hunk 2]
```diff
@@ -178,23 +180,37 @@ public final boolean perform(final AbstractBuild<?, ?> build, final Launcher lau
...
+        }
+        if (customComment != null) {
```
Comment (jjx) [inline src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java]: Can you just move this logic inside the try? No point in executing it if we hit one of the exceptions. Also, you can use CommonUtils.isBlank() to check for null or empty strings

File: src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java (hunk 1 above)
Comment (sectioneight) [inline src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java]: Done

File: src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java (hunk 2 above)
Comment (sectioneight) [inline src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java]: Good call. Done to both.

File: src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java [hunk 3]
```diff
@@ -232,6 +248,31 @@ private String getCoverageComment(Ratio lineCoverage, UberallsClient uberalls, D
...
+        FilePath workspace = build.getWorkspace();
+        FilePath[] src = workspace.list(commentFile);
+        if (src.length == 0) {
+            logger.println("[comment-file] no files found by path: '" + commentFile + "'");
+            return null;
+        }
+        if (src.length > 1) {
+            logger.println("[comment-file] Found multiple matches. Reading first only.");
+        }
+        return src[0].readToString();
```
Comment (sectioneight) [inline src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java]: not sure if i should use a stream here and/or limit the size somehow

File: src/main/java/com/uber/jenkins/phabricator/uberalls/UberallsClient.java [hunk 4]
```diff
@@ -79,9 +81,10 @@ public boolean recordCoverage(String currentSHA, String branch, CodeCoverageMetr
         return false;
     }
 
-    public UberallsClient(String baseURL, EnvVars environment) {
+    public UberallsClient(String baseURL, EnvVars environment, PrintStream logger) {
         this.baseURL = baseURL;
         this.environment = environment;
+        this.logger = logger;
```
Comment (jjx) [inline src/main/java/com/uber/jenkins/phabricator/uberalls/UberallsClient.java]: <3

File: src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java [hunk 5]
```diff
@@ -232,6 +248,31 @@ private String getCoverageComment(Ratio lineCoverage, UberallsClient uberalls, D
...
+        }
+
+        FilePath workspace = build.getWorkspace();
+        FilePath[] src = workspace.list(commentFile);
+        if (src.length == 0) {
+            logger.println("[comment-file] no files found by path: '" + commentFile + "'");
+            return null;
+        }
+        if (src.length > 1) {
+            logger.println("[comment-file] Found multiple matches. Reading first only.");
```
Comment (jjx) [inline src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java]: Is this code reachable? Does the plugin need to take in some kind of wildcard in order for us to get here?

Comment (sectioneight) [inline src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java]: somebody could put in a wildcard, in which case the list() call would return multiple matches. i'm pretty obviously bad at java so if there is a better API to use i'm all for it (although i imagine in the future we may want to recursively search for .comment files like we do for junit)

File: src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java (hunk 3 above)
Comment (jjx) [inline src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java]: Good point. Is there a limit to the arc comment size?

Comment (sectioneight) [inline src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java]: not currently. i guess i could trim the string to say, 1000 characters? 10,000?

Comment (jjx) [inline src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java]: Lorem ipsum with 1000 bytes is 1+ paragraphs, that should be more than enough IMO. Let's make the limit a param in this file so it's easy to update if we ever need to

**Comment by sectioneight** [comment]
@jjx I'm having problems with JSON-escaping:

```
[workspace] $ sh -c 'echo '{"silent":false,"message":"Build has FAILED\\n\\n```\\nsrc\\/code.uber.internal\\/go-common.git\\/x\\/errors\\/multi.go:18:1: comment on exported method Multierr.Set should be of the form \\\"Set ...\\\"\\n\\n```\\n http:\\/\\/localhost:8080\\/jenkins\\/job\\/do-thing\\/38\\/ for more details.","revision_id":"1638","action":"reject"}' | arc call-conduit differential.createcomment'
[phabricator] Unable to parse JSON from response: Usage Exception: Provide method parameters on stdin as a JSON blob.
```
It looks like it's having trouble with built-in quotes, because this works:
```
[workspace] $ sh -c 'echo '{"silent":false,"message":"Build has FAILED\\n\\n```\\nsrc\\/code.uber.internal\\/go-common.git\\/x\\/errors\\/multi.go:18:1: comment on exported method Multierr.Set should be of the form Set ...\\n\\n```\\n http:\\/\\/localhost:8080\\/jenkins\\/job\\/do-thing\\/37\\/ for more details.","revision_id":"1638","action":"none"}' | arc call-conduit differential.createcomment'
```

**Comment by sectioneight** [comment]
Fixed the double-escaping problem

**Comment by jjx** [comment]
lgtm

**Pull request merged**